### PR TITLE
Move distribution of unallocated users to adhoc task

### LINKED ISSUE #263: Distribution of unallocated users as background process
Currently, the distribution of unallocated users will be done directly in the same request that the user triggers by hitting the button. Sometimes, with many students to allocate and group restrictions active this can take some moments. To avoid that the user runs into a timeout or considers the request to be failed it would be safer to let the calculation being done in the background by a separate process.

1. option: Clicking the button will add the distribution task to be picked up by the same scheduled task currently responsible for the normal distribution algorithm.
2. option: As far as I understand the only purpose of using a scheduled task instead of an ad hoc task is that it is possible to make the algorithm run after the voting phase has ended so the trainer does not have to trigger it manually. The distribution of unallocated users does not have that. So it maybe would be sufficient to just use a simple ad-hoc task for that.
3. option: Unify both options and refactor the current scheduled task to just be a simple ad-hoc task. However that means (from my understanding) that the trainer will have to trigger the algorithm manually which would be fine for me, because most of my users do not really understand that the distribution happens automatically anyway :-)

BTW: Should it be made configurable that the algorithm is being run in the background or should this just be forced?

Any feedback welcome!

### PR DESCRIPTION
**Pull request by PhMemmel**
Closes #263

### CODE REVIEW

**Review by Laur0r — COMMENTED** [review]
Thank you very much for your contribution and fast implementation. I only found one small issue, otherwise it looks great :)

File: renderer.php [hunk 1]
```diff
@@ -782,4 +791,15 @@ private function add_table_row_tuple(html_table $table, $first, $second) {
...
     }
+
+    /**
+     * Method to check if an adhoc task for distributing unallocated users has already been queued.
+     *
+     * @return bool true if an adhoc task for the current course module can be found, false otherwise
+     */
+    private function is_distribution_of_unallocated_users_running(int $coursemoduleid): bool {
+        $queuedtasks = \core\task\manager::get_adhoc_tasks(\mod_ratingallocate\task\distribute_unallocated_task::class);
+        $taskofcurrentmodule = array_filter($queuedtasks, fn($task) => $task->get_custom_data()->cmid === $coursemoduleid);
```
[the remark this answers is not in the document]
Comment (PhMemmel) [reply renderer.php]: Thank you. As I'm not a big fan of type juggling I used `intval` to fix this.

[the remark this answers is not in the document]
Comment (Laur0r) [reply renderer.php]: That is of course also a good solution!

**Review by Laur0r — APPROVED** [review]
Great, thank you!

**Pull request merged**